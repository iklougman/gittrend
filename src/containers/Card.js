import React, { useState } from "react";
import PropTypes from "prop-types";
import { Button, Tooltip, Card as AntCard } from "antd";
import { StarOutlined, StarFilled } from "@ant-design/icons";
import { isStarred, starRepo, unStarRepo } from "../utils";
import Meta from "antd/lib/card/Meta";

// Component
const Card = ({ repo = {} }) => {
  const { owner = {} } = repo;
  const {
    id = null,
    full_name = "No name",
    html_url = "",
    stargazers_count = 0,
    language = "-",
    description = "No description",
  } = repo;
  const [userStarred, setUserStarred] = useState(false);
  const hasStar = isStarred(id);
  const handleStarRepo = () => {
    if (isStarred(id)) {
      unStarRepo(id);
    } else {
      starRepo(id);
    }
    setUserStarred(!userStarred);
  };
  const stars = stargazers_count + (isStarred(id) ? 1 : 0);
  return (
    <div>
      <AntCard
        data-testid="repo-card"
        className="repo-card"
        cover={<img alt="example" src={owner.avatar_url} />}
      >
        {Object.keys(repo).length > 0 && (
          <Tooltip title={hasStar ? "unstar" : "give a star"}>
            <Button
              block
              data-testid="star-repo-btn"
              type={hasStar ? "primary" : "secondary"}
              onClick={handleStarRepo}
              icon={hasStar ? <StarFilled /> : <StarOutlined />}
            >
              {stars} Stars
            </Button>
          </Tooltip>
        )}
        <h2>
          <a href={html_url} alt={html_url}>
            {full_name}
          </a>
        </h2>
        <Tooltip placement="bottom" title={description || "-"}>
          <Meta title={description || "-"} />
        </Tooltip>
        <br />
        <Meta description={`Language: ${language}` || "-"} />
        <br />
        <a alt={`star number ${stars}`} href={`${html_url}/stargazers`}>
          <p>Show Stargazers</p>
        </a>
      </AntCard>
    </div>
  );
};

Card.propTypes = {
  repo: PropTypes.object.isRequired,
};

export default Card;
