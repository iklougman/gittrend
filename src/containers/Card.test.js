import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import Card from "./Card";
import { mockRepoOne } from "./fixtures/github_repository_mock";

describe("<Card />", () => {
  describe("Given an empty repository object", () => {
    it("Shows ean empty card with default data ", async () => {
      const mockEmptyRepository = {};
      const { asFragment } = render(<Card repo={mockEmptyRepository} />);

      expect(asFragment()).toMatchSnapshot();
      screen.getByText("No description");
    });
  });

  describe("Given correct github repository data object:", () => {
    beforeEach(() => {
      localStorage.clear();
    });
    it("Shows ean empty card with default data ", async () => {
      const { asFragment } = render(<Card repo={mockRepoOne} />);

      screen.getByText(/hivemind/);
      screen.getByText(/253/);

      screen.getByText(
        "For creating distributed jobs using AWS Lambda functions"
      );
      expect(asFragment()).toMatchSnapshot();
    });
    describe("When clicking on the 'Star Repo' button", () => {
      it("Saves in the localStorage the 'id' of the starred repository", () => {
        render(<Card repo={mockRepoOne} />);
        expect(localStorage.getItem(101689336)).toBeNull();
        fireEvent.click(screen.getByTestId("star-repo-btn"));

        expect(localStorage.getItem(101689336)).not.toBeNull();
      });
    });
    describe("When clicking on the 'Star Repo' button of an already starred repository", () => {
      it("Removes from the localStorage the 'id' of the starred repository", async () => {
        render(<Card repo={mockRepoOne} />);
        fireEvent.click(screen.getByTestId("star-repo-btn"));
        expect(localStorage.getItem(101689336)).not.toBeNull();
        fireEvent.click(screen.getByTestId("star-repo-btn"));
        expect(localStorage.getItem(101689336)).toBeNull();
      });
    });
  });
});
