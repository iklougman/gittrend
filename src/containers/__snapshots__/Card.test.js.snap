// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Given an empty repository object Shows ean empty card with default data  1`] = `
<DocumentFragment>
  <div
    class="ant-card repo-card ant-card-bordered"
    data-testid="repo-card"
  >
    <div
      class="ant-card-cover"
    >
      <img
        alt="example"
      />
    </div>
    <div
      class="ant-card-body"
    >
      <h2>
        <a
          alt=""
          href=""
        >
          No name
        </a>
      </h2>
      <div
        class="ant-card-meta"
      >
        <div
          class="ant-card-meta-detail"
        >
          <div
            class="ant-card-meta-title"
          >
            No description
          </div>
        </div>
      </div>
      <br />
      <div
        class="ant-card-meta"
      >
        <div
          class="ant-card-meta-detail"
        >
          <div
            class="ant-card-meta-description"
          >
            Language: -
          </div>
        </div>
      </div>
      <br />
      <a
        alt="star number 0"
        href="/stargazers"
      >
        <p>
          Show Stargazers
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Card /> Given correct github repository data object: Shows ean empty card with default data  1`] = `
<DocumentFragment>
  <div
    class="ant-card repo-card ant-card-bordered"
    data-testid="repo-card"
  >
    <div
      class="ant-card-cover"
    >
      <img
        alt="example"
        src="https://avatars2.githubusercontent.com/u/5302528?v=4"
      />
    </div>
    <div
      class="ant-card-body"
    >
      <button
        class="ant-btn ant-btn-secondary ant-btn-block"
        data-testid="star-repo-btn"
        type="button"
      >
        <span
          aria-label="star"
          class="anticon anticon-star"
          role="img"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="star"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 00.6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0046.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
            />
          </svg>
        </span>
        <span>
          253 Stars
        </span>
      </button>
      <h2>
        <a
          alt="https://github.com/littlstar/hivemind"
          href="https://github.com/littlstar/hivemind"
        >
          littlstar/hivemind
        </a>
      </h2>
      <div
        class="ant-card-meta"
      >
        <div
          class="ant-card-meta-detail"
        >
          <div
            class="ant-card-meta-title"
          >
            For creating distributed jobs using AWS Lambda functions
          </div>
        </div>
      </div>
      <br />
      <div
        class="ant-card-meta"
      >
        <div
          class="ant-card-meta-detail"
        >
          <div
            class="ant-card-meta-description"
          >
            Language: JavaScript
          </div>
        </div>
      </div>
      <br />
      <a
        alt="star number 253"
        href="https://github.com/littlstar/hivemind/stargazers"
      >
        <p>
          Show Stargazers
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
